src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -11,32 +11,62 @@ namespace System.Net
 {
     internal class IPAddressParser
     {
+
+        internal static unsafe bool ContainsColon(char* ipStringPtr, int ipStringPtrLength)
+        {
+            for (var i = 0; i < ipStringPtrLength; i++)
+            {
+                if (ipStringPtr[i] == ':')
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
         internal static unsafe IPAddress Parse(string ipString, bool tryParse)
         {
-            if (ipString == null)
+            fixed (char* ipStringPtr = ipString)
+            {
+                return Parse(ipStringPtr, ipString.Length, tryParse);
+            }
+        }
+
+        internal static unsafe IPAddress Parse(ReadOnlySpan<char> ipSpan, bool tryParse)
+        {
+            fixed (char* ipStringPtr = &ipSpan.DangerousGetPinnableReference())
+            {
+                return Parse(ipStringPtr, ipSpan.Length, tryParse);
+            }
+
+        }
+   
+        internal static unsafe IPAddress Parse(char* ipStringPtr, int ipStringPtrLength, bool tryParse)
+        {
+            if (ipStringPtr == null)
             {
                 if (tryParse)
                 {
                     return null;
                 }
-                throw new ArgumentNullException(nameof(ipString));
+                throw new ArgumentNullException(nameof(ipStringPtr));
             }
 
-            if (ipString.IndexOf(':') != -1)
+            if (ContainsColon(ipStringPtr, ipStringPtrLength))
```

Comment:
Especially in networking-related code, we want to avoid unnecessary unsafe code, so it'd be great to go the route of using spans as much as possible rather than introducing additional pointer-based code.  For now I'd minimize changing pointer-based code to spans, as it could have more perf impact right now than we want to accept, but when converting string-based methods to something else, let's prefer spans over pointers.